REGRESSION(294049@main): Comic-Con website may show menu content partly offscreen

graouts opened the pull request:
#### 8fd5ab7bf0260ad28d50e10b8a66a0b1bf9eac07
<pre>
REGRESSION(294049@main): Comic-Con website may show menu content partly offscreen
<a href="https://bugs.webkit.org/show_bug.cgi?id=295184">https://bugs.webkit.org/show_bug.cgi?id=295184</a>
<a href="https://rdar.apple.com/152375667">rdar://152375667</a>

Reviewed by Simon Fraser.

Prior to 294049@main, when the current time would be requested from `AnimationTimelinesController`,
most likely through `DocumentTimeline::currentTime()`, we would either use the cached current time,
if available, or compute a new one. In the case where we would compute a new one, we would also
enqueue a task (using `EventLoop::queueTask()`) to clear it after the current run loop had completed,
ensuring that any code ran within that loop would use the same current time.

That changed in 294049@main where we made it so that the document timeline&apos;s current time was cached
throughout the duration of the animation frame, matching the behavior of Chrome and Firefox. In effect,
this meant that that current time was updated every 16ms or so (assuming a display update cadence of 60Hz).

However, we neglected to preserve the mechanism to clear that cached current time, instead relying on
the fact that animations would cause animations to be updated and for that current time to be cached
as the page rendering updated at 60Hz.

The Comic-Con website has a menu which, in a narrow window or on a narrow display such as an iPhone,
animates an element which is the width of the window over 150ms from right to left, taking it from
being completely offscreen to completely onscreen. That element is contained within an element that
has `overflow-x: hidden`, meaning that the animated element animates from being completely in the
overflow part of its container to. That element has a title wrapped in a &lt;button&gt; element which has
its `focus()` method called on it 50ms into the animated transition using a `setTimeout()` call.

Sometimes, this transition would appear to end up partly offscreen on the left of the window. The
reason this would happen is that the animated transition in question was performed using the `transform`
property, meaning that acceleration was used. In that scenario, we don&apos;t schedule animation updates
while the transition is in-flight, letting Core Animation update the animated time and interpolate
the transform property of the composited layer backing the target element. This meant that we would have
no chance, within WebCore, to update the transition&apos;s timeline&apos;s (ie. the document timeline) current time
while the transition would be in-flight.

However, when `focus()` is called on an element within a scrollable container, WebCore will scroll
that element into view if it deems this necessary (ie. it&apos;s partially invisible). To achieve this,
the method `RenderLayerScrollableArea::updateScrollPositionForScrollIntoView()` is called. This method
will eventually call into `RenderElement::absoluteAnchorRectWithScrollMargin()` which itself will call
into `KeyframeEffect::getAnimatedStyle()`. The purpose of that last method is to compute the animated
style for this keyframe effect on demand when the target element&apos;s renderer&apos;s style has not been updated
during the traditional process of updating animations during page rendering updates, as is the case
for accelerated animations.

However, since we would fail to update the document timeline&apos;s current time past the start of the
transition, we would compute the bounds incorrectly and assume the element was contained in the overflow
area. This would lead to the container with `overflow-x: hidden` to scroll to reveal the `&lt;button&gt;`
element.

In this patch we address this particular issue by adding a way to *not* use the cached current time
when computing an animation&apos;s current time, which relies on the document timeline&apos;s current time. We
elect to do so in the case where `KeyframeEffect::getAnimatedStyle()` is called for an effect that
is running with acceleration.

Note that this doesn&apos;t fix the problem of not clearing the cached current time in the absence of
repeated page rendering updates. This will be addressed separately in bug 295177.

Finally, to ensure we don&apos;t encounter this issue going forward, we add a test that replicates the
behavior of the Comic-Con website. This test would reliably fail prior to this patch.

* LayoutTests/webanimations/focus-accelerated-animated-element-initially-in-overflow-expected.txt: Added.
* LayoutTests/webanimations/focus-accelerated-animated-element-initially-in-overflow.html: Added.
* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::AnimationEffect::resolutionData const):
(WebCore::AnimationEffect::getComputedTiming):
* Source/WebCore/animation/AnimationEffect.h:
* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::AnimationTimeline::currentTime):
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::currentTime):
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::currentTime):
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::getAnimatedStyle):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::currentTime):
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::currentTime const):
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/animation/WebAnimationTypes.h:

Canonical link: <a href="https://commits.webkit.org/296833@main">https://commits.webkit.org/296833@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef5683dc8d0958bfabbb02767c72186ff9751ed3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109744 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29402 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19832 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/115765 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/59978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/111707 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30080 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37990 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/115765 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/59978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112692 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/23968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/98833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/115765 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/23348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/16978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/59559 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/93342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/17019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118557 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36783 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/27245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/92401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37156 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/95095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/92223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/37189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/14932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32611 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17710 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36678 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42148 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/36338 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39680 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38047 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on LayoutTests/webanimations/focus-accelerated-animated-element-initially-in-overflow.html:
```diff
@@ -0,0 +1,81 @@
+</script>
+
+</body>
+</html>
```
nit: add empty newline to EOF

smfr commented on LayoutTests/webanimations/focus-accelerated-animated-element-initially-in-overflow.html:
```diff
@@ -0,0 +1,81 @@
+<html>
```
`DOCTYPE`?

graouts committed:
REGRESSION(294049@main): Comic-Con website may show menu content partly offscreen
https://bugs.webkit.org/show_bug.cgi?id=295184
rdar://152375667

Reviewed by Simon Fraser.

Prior to 294049@main, when the current time would be requested from `AnimationTimelinesController`,
most likely through `DocumentTimeline::currentTime()`, we would either use the cached current time,
if available, or compute a new one. In the case where we would compute a new one, we would also
enqueue a task (using `EventLoop::queueTask()`) to clear it after the current run loop had completed,
ensuring that any code ran within that loop would use the same current time.

That changed in 294049@main where we made it so that the document timeline's current time was cached
throughout the duration of the animation frame, matching the behavior of Chrome and Firefox. In effect,
this meant that that current time was updated every 16ms or so (assuming a display update cadence of 60Hz).

However, we neglected to preserve the mechanism to clear that cached current time, instead relying on
the fact that animations would cause animations to be updated and for that current time to be cached
as the page rendering updated at 60Hz.

The Comic-Con website has a menu which, in a narrow window or on a narrow display such as an iPhone,
animates an element which is the width of the window over 150ms from right to left, taking it from
being completely offscreen to completely onscreen. That element is contained within an element that
has `overflow-x: hidden`, meaning that the animated element animates from being completely in the
overflow part of its container to. That element has a title wrapped in a <button> element which has
its `focus()` method called on it 50ms into the animated transition using a `setTimeout()` call.

Sometimes, this transition would appear to end up partly offscreen on the left of the window. The
reason this would happen is that the animated transition in question was performed using the `transform`
property, meaning that acceleration was used. In that scenario, we don't schedule animation updates
while the transition is in-flight, letting Core Animation update the animated time and interpolate
the transform property of the composited layer backing the target element. This meant that we would have
no chance, within WebCore, to update the transition's timeline's (ie. the document timeline) current time
while the transition would be in-flight.

However, when `focus()` is called on an element within a scrollable container, WebCore will scroll
that element into view if it deems this necessary (ie. it's partially invisible). To achieve this,
the method `RenderLayerScrollableArea::updateScrollPositionForScrollIntoView()` is called. This method
will eventually call into `RenderElement::absoluteAnchorRectWithScrollMargin()` which itself will call
into `KeyframeEffect::getAnimatedStyle()`. The purpose of that last method is to compute the animated
style for this keyframe effect on demand when the target element's renderer's style has not been updated
during the traditional process of updating animations during page rendering updates, as is the case
for accelerated animations.

However, since we would fail to update the document timeline's current time past the start of the
transition, we would compute the bounds incorrectly and assume the element was contained in the overflow
area. This would lead to the container with `overflow-x: hidden` to scroll to reveal the `<button>`
element.

In this patch we address this particular issue by adding a way to *not* use the cached current time
when computing an animation's current time, which relies on the document timeline's current time. We
elect to do so in the case where `KeyframeEffect::getAnimatedStyle()` is called for an effect that
is running with acceleration.

Note that this doesn't fix the problem of not clearing the cached current time in the absence of
repeated page rendering updates. This will be addressed separately in bug 295177.

Finally, to ensure we don't encounter this issue going forward, we add a test that replicates the
behavior of the Comic-Con website. This test would reliably fail prior to this patch.

* LayoutTests/webanimations/focus-accelerated-animated-element-initially-in-overflow-expected.txt: Added.
* LayoutTests/webanimations/focus-accelerated-animated-element-initially-in-overflow.html: Added.
* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::AnimationEffect::resolutionData const):
(WebCore::AnimationEffect::getComputedTiming):
* Source/WebCore/animation/AnimationEffect.h:
* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::AnimationTimeline::currentTime):
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::currentTime):
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::currentTime):
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::getAnimatedStyle):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::currentTime):
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::currentTime const):
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/animation/WebAnimationTypes.h:

Canonical link: https://commits.webkit.org/296833@main
